enhancements/insights/on-demand-data-gathering.md

```diff
@@ -0,0 +1,340 @@
+This new CRD will build on and extend the existing `GatherConfig` (see the [Insights config API enhancement](../insights/insights-config-api.md)). The idea is that with each new CR instance of this resource, new data gathering will happen. 
+
+The proposal for the new `datagather.insights.openshift.io` resource is:
+
+```go
+type DataGather struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec holds user settable values for configuration
+	// +kubebuilder:validation:Required
+	Spec DataGatherSpec `json:"spec"`
+	// status holds observed values from the cluster. They may not be overridden.
+	// +optional
+	Status DataGatherStatus `json:"status"`
+}
+
+type DataGatherSpec struct {
+	// dataPolicy allows user to enable additional global obfuscation of the IP addresses and base domain
+	// in the Insights archive data. Valid values are "None" and "ObfuscateNetworking".
+	// When set to None the data is not obfuscated.
+	// When set to ObfuscateNetworking the IP addresses and the cluster domain name are obfuscated.
+	// When omitted, this means no opinion and the platform is left to choose a reasonable default, which is subject to change over time.
+	// The current default is None.
+	// +optional
+	DataPolicy DataPolicy `json:"dataPolicy"`
+	// disabledGatherers is a list of gatherers to be excluded from the gathering. All the gatherers can be disabled by providing "all" value.
+	// If all the gatherers are disabled, the Insights operator does not gather any data.
+	// The particular gatherers IDs can be found at https://github.com/openshift/insights-operator/blob/master/docs/gathered-data.md.
+	// Run the following command to get the names of last active gatherers:
+	// "oc get insightsoperators.operator.openshift.io cluster -o json | jq '.status.gatherStatus.gatherers[].name'"
+	// An example of disabling gatherers looks like this: `disabledGatherers: ["clusterconfig/machine_configs", "workloads/workload_info"]`
+	// +optional
+	DisabledGatherers []string `json:"disabledGatherers"`
+}
+
+const (
+	// No data obfuscation
+	NoPolicy DataPolicy = "None"
+	// IP addresses and cluster domain name are obfuscated
+	ObfuscateNetworking DataPolicy = "ObfuscateNetworking"
+	// Data gathering is running
+	Running DataGatherState = "Running"
+	// Data gathering is completed
+	Completed DataGatherState = "Completed"
+	// Data gathering failed
+	Failed DataGatherState = "Failed"
+	// Data gathering is pending
+	Pending DataGatherState = "Pending"
+)
+
+// dataPolicy declares valid data policy types
+// +kubebuilder:validation:Enum="";None;ObfuscateNetworking
+type DataPolicy string
+
+// dataGatherState declares valid gathering state types
+// +kubebuilder:validation:Enum=Running;Completed;Failed;Pending
+type DataGatherState string
+
+type DataGatherStatus struct {
+	// conditions provide details on the status of the gatherer job.
+	// +listType=atomic
```

Comment:
//	    // +patchMergeKey=type
//	    // +patchStrategy=merge
//	    // +listType=map
//	    // +listMapKey=type

Reply:
Can you please explain why? What would be different to https://github.com/openshift/api/blob/master/operator/v1/types_insights.go#L121 ?

Reply:
I think tbh the other API should be evolved and we should use the patch/list type keys, Conditions slices are pretty much universally treated as a list map, as it allows multiple actors to act on the list at once. For example, you can have the business logic of your operator broken down into multiple controllers and they can each own some subset of the conditions, and not need to worry about stomping each others conditions.

I know we talked about that not necessarily being needed by the insights, but there isn't really any harm in changing it I don't think

Reply:
OK. I've updated here but I am having some troubles with the update target in the api repo..